File: opencode-web-skill/opencode_wrapper/manager.py
import time
import requests
import queue
import logging
from threading import Thread, Event
from .config import OPENCODE_URL, PROJECT_ROOT
from .worker import Worker

logger = logging.getLogger("OpencodeManager")

class SessionManager(Thread):
    def __init__(self, session_id):
        super().__init__()
        self.session_id = session_id
        self.input_queue = queue.Queue()
        self.worker = None
        self.state = "IDLE" 
        self.latest_response = None
        self.questions = []
        self._stop_event = Event()
        self.last_activity = 0
        self.daemon = True # Run as daemon thread
        
    def run(self):
        while not self._stop_event.is_set():
            try:
                # 1. Process Input
                req = self.input_queue.get(timeout=1.0)
                self._handle_request(req)
            except queue.Empty:
                pass
            
            # 2. Check Worker Status
            if self.worker and not self.worker.is_alive():
                # Worker finished
                self.worker = None
                if self.state == "BUSY":
                    self.state = "IDLE"

            # 3. Poll Questions (Throttle to 2s)
            if time.time() - self.last_activity > 2.0:
                 self._poll_questions()
                 self.last_activity = time.time()

    def submit_request(self, req):
        self.input_queue.put(req)

    def get_snapshot(self):
        return {
            "state": self.state,
            "session_id": self.session_id,
            "latest_response": self.latest_response,
            "questions": self.questions
        }

    def _handle_request(self, req):
        r_type = req.get("type")
        payload = req.get("payload")
        
        if r_type in ["PROMPT", "COMMAND"]:
            if self.worker and self.worker.is_alive():
                # We ignore new prompts if busy (Client should check state first)
                logger.warning(f"Session {self.session_id} is busy. Ignoring {r_type}.")
                return
            
            self.state = "BUSY"
            self.latest_response = None # Clear previous
            endpoint = "command" if r_type == "COMMAND" else "message"
            self.worker = Worker(self.session_id, payload, self.on_worker_done, endpoint=endpoint)
            self.worker.start()
            
        elif r_type == "ANSWER":
            # Handle Answer immediately (unblocks agent)
            try:
                url = f"{OPENCODE_URL}/question/{payload['requestID']}/reply"
                headers = {
                    "Content-Type": "application/json",
                    "x-opencode-directory": str(PROJECT_ROOT)
                }
                # Wrap answer in list if needed or assume Client formatted it
                # Wrapper protocol: Client sends formatted payload { "answers": [...] }
                # The payload here is exactly what we send to API? 
                # Let's assume payload has "answers" key.
                requests.post(url, json={"answers": payload["answers"]}, headers=headers)
                
                # Optimistically remove question
                self.questions = [q for q in self.questions if q["id"] != payload["requestID"]]
                
                # Update State
                if not self.questions:
                    if self.worker and self.worker.is_alive():
                        self.state = "BUSY"
                    else:
                        self.state = "IDLE"
            except Exception as e:
                logger.error(f"Answer failed: {e}")

        elif r_type == "FIX":
            try:
                # 1. Abort
                logger.info(f"Aborting session {self.session_id}...")
                requests.post(f"{OPENCODE_URL}/session/{self.session_id}/abort", 
                              json={}, 
                              headers={"x-opencode-directory": str(PROJECT_ROOT)})
                
                # 2. Reset Worker (if possible, start new)
                # We can't easily kill the old thread, but we can ignore it.
                # However, sending 'continue' message might fail if old request is still pending?
                # Abort should make the old request return.
                # Let's hope the old worker finishes quickly.
                
                # 3. Send Continue
                logger.info(f"Sending continue to session {self.session_id}...")
                self.latest_response = None
                self.state = "BUSY"
                
                payload = {
                    "agent": "sisyphus", # Default fallback
                    "model": {"providerID": "zai-coding-plan", "modelID": "glm-4.7"}, # Default fallback
                    "parts": [{"type": "text", "text": "continue"}]
                }
                # Use default model/agent if payload missing? Client should provide them ideally.
                # Assuming generic continue is fine with defaults.
                
                # We overwrite self.worker. The old thread becomes orphaned and eventually dies.
                self.worker = Worker(self.session_id, payload, self.on_worker_done, endpoint="message")
                self.worker.start()
                
            except Exception as e:
                logger.error(f"Fix failed: {e}")
                self.state = "IDLE"
                self.latest_response = {"result": None, "error": f"Fix failed: {e}"}

    def on_worker_done(self, result, error):
        self.latest_response = {"result": result, "error": error}
        logger.info(f"Worker finished for {self.session_id}")

    def _poll_questions(self):
        try:
            url = f"{OPENCODE_URL}/question"
            headers = {"x-opencode-directory": str(PROJECT_ROOT)}
            resp = requests.get(url, headers=headers, timeout=5)
            if resp.status_code == 200:
                resp_json = resp.json()
                data = resp_json if isinstance(resp_json, list) else resp_json.get('data', [])
                
                # Filter for this session
                self.questions = [q for q in data if q.get('sessionID') == self.session_id]
                
                if self.questions:
                    self.state = "WAITING_FOR_INPUT"
                elif self.state == "WAITING_FOR_INPUT":
                    # If invalid/gone, and worker is running -> BUSY
                    if self.worker and self.worker.is_alive():
                        self.state = "BUSY"
                    else:
                        self.state = "IDLE"
        except Exception as e:
            logger.error(f"Poll Questions Error: {e}")

    def stop(self):
        self._stop_event.set()

File: opencode-web-skill/opencode_wrapper_go/internal/manager/manager.go
package manager

import (
	"log"
	"time"

	"opencode_wrapper/internal/api"
	"opencode_wrapper/internal/config"
)

type State string

const (
	StateIdle            State = "IDLE"
	StateBusy            State = "BUSY"
	StateWaitingForInput State = "WAITING_FOR_INPUT"
)

type SessionManager struct {
	SessionID      string
	State          State
	LatestResponse interface{}
	Questions      []api.Question

	inputChan chan Request
	stopChan  chan struct{}
	client    *api.Client

	// Worker tracking
	workerDoneChan chan workerResult
	isWorkerBusy   bool
	taskStartTime  time.Time
	lastActivity   time.Time
}

type Request struct {
	Type       string
	Payload    interface{}
	ResultChan chan error // Optional, for sync acknowledgement
}

type workerResult struct {
	Result interface{}
	Error  error
}

func NewSessionManager(sessionID string, workingDir string) *SessionManager {
	return &SessionManager{
		SessionID:      sessionID,
		State:          StateIdle,
		inputChan:      make(chan Request, 10),
		stopChan:       make(chan struct{}),
		workerDoneChan: make(chan workerResult, 1),
		client:         api.NewClient(workingDir),
		lastActivity:   time.Now(),
	}
}

func (sm *SessionManager) Start() {
	go sm.loop()
}

func (sm *SessionManager) UpdateWorkingDir(workingDir string) {
	sm.client = api.NewClient(workingDir)
}

func (sm *SessionManager) Stop() {
	close(sm.stopChan)
}

func (sm *SessionManager) SubmitRequest(req Request) {
	sm.inputChan <- req
}

func (sm *SessionManager) GetSnapshot() map[string]interface{} {
	// Note: accessing fields without lock is technically racy but for a simple snapshot it might be okay.
	// Ideally we should use a command to get snapshot or a mutex.
	// For simplicity, let's use a mutex-less approach assuming single writer (the loop) and readers.
	// Or better, send a request to get snapshot? No, that blocks.
	// Let's just read.
	return map[string]interface{}{
		"state":           sm.State,
		"session_id":      sm.SessionID,
		"latest_response": sm.LatestResponse,
		"questions":       sm.Questions,
	}
}

func (sm *SessionManager) loop() {
	ticker := time.NewTicker(config.PollInterval)
	defer ticker.Stop()

	for {
		select {
		case <-sm.stopChan:
			return

		case req := <-sm.inputChan:
			sm.handleRequest(req)

		case res := <-sm.workerDoneChan:
			sm.handleWorkerDone(res)

		case <-ticker.C:
			sm.pollQuestions()
			sm.checkAutoFix()
		}
	}
}

func (sm *SessionManager) handleRequest(req Request) {
	log.Printf("Handling request type: %s", req.Type)
	if req.ResultChan != nil {
		defer close(req.ResultChan)
	}

	switch req.Type {
	case "PROMPT", "COMMAND":
		if sm.isWorkerBusy {
			// Already checked by Daemon, but double check
			if req.ResultChan != nil {
				req.ResultChan <- nil
			}
			return
		}

		sm.State = StateBusy
		sm.LatestResponse = nil
		sm.taskStartTime = time.Now()
		sm.isWorkerBusy = true

		log.Printf("Starting worker for PROMPT/COMMAND...")
		go sm.runWorker(req)

	case "ANSWER":
		payload, ok := req.Payload.(api.AnswerRequest)
		if ok {
			if err := sm.client.AnswerQuestion(payload); err != nil {
				log.Printf("Answer failed: %v", err)
			} else {
				// Optimistically remove question
				newQuestions := []api.Question{}
				for _, q := range sm.Questions {
					if q.ID != payload.RequestID {
						newQuestions = append(newQuestions, q)
					}
				}
				sm.Questions = newQuestions

				if len(sm.Questions) == 0 {
					if sm.isWorkerBusy {
						sm.State = StateBusy
						sm.taskStartTime = time.Now() // Reset timeout
					} else {
						sm.State = StateIdle
					}
				}
			}
		}

	case "FIX":
		sm.performFix()
	}

	if req.ResultChan != nil {
		req.ResultChan <- nil
	}
}

func (sm *SessionManager) runWorker(req Request) {
	var res interface{}
	var err error

	if req.Type == "COMMAND" {
		cmdReq, _ := req.Payload.(api.CommandRequest)
		res, err = sm.client.SendCommand(sm.SessionID, cmdReq)
	} else {
		promptReq, _ := req.Payload.(api.PromptRequest)
		res, err = sm.client.SendPrompt(sm.SessionID, promptReq)
	}

	sm.workerDoneChan <- workerResult{Result: res, Error: err}
}

func (sm *SessionManager) handleWorkerDone(res workerResult) {
	sm.isWorkerBusy = false

	if res.Error != nil {
		sm.LatestResponse = map[string]interface{}{"error": res.Error.Error()}
	} else {
		sm.LatestResponse = map[string]interface{}{"result": res.Result}
	}

	if len(sm.Questions) > 0 {
		sm.State = StateWaitingForInput
	} else {
		sm.State = StateIdle
	}
}

func (sm *SessionManager) pollQuestions() {
	questions, err := sm.client.GetQuestions()
	if err != nil {
		log.Printf("Poll error: %v", err)
		return
	}

	// Filter for this session
	sessionQuestions := []api.Question{}
	for _, q := range questions {
		if q.SessionID == sm.SessionID {
			sessionQuestions = append(sessionQuestions, q)
		}
	}
	sm.Questions = sessionQuestions

	if len(sm.Questions) > 0 {
		sm.State = StateWaitingForInput
	} else if sm.State == StateWaitingForInput {
		if sm.isWorkerBusy {
			sm.State = StateBusy
		} else {
			sm.State = StateIdle
		}
	}
}

func (sm *SessionManager) checkAutoFix() {
	if len(sm.Questions) > 0 {
		return
	}

	if sm.State == StateBusy && sm.isWorkerBusy {
		if time.Since(sm.taskStartTime) > config.AutoFixTimeout {
			log.Printf("Session %s exceeded timeout. Triggering Auto-Fix.", sm.SessionID)
			// We need to trigger FIX. We can send a request to ourselves?
			// Or just call performFix directly?
			// performFix is blocking-ish, better to schedule it
			go func() {
				sm.inputChan <- Request{Type: "FIX"}
			}()
		}
	}
}

func (sm *SessionManager) performFix() {
	log.Printf("Performing FIX for session %s...", sm.SessionID)

	// 1. Abort
	_ = sm.client.AbortSession(sm.SessionID)

	// Wait
	time.Sleep(3 * time.Second)

	// 2. Reset Worker state?
	// The previous worker might still return, but we overwrite it.

	// 3. Send Continue
	sm.isWorkerBusy = true
	sm.State = StateBusy
	sm.taskStartTime = time.Now()
	sm.LatestResponse = nil

	req := api.PromptRequest{
		Agent: "sisyphus",
		Model: api.ModelDetails{ProviderID: "zai-coding-plan", ModelID: "glm-5"},
		Parts: []api.Part{{Type: "text", Text: "continue"}},
	}

	go func() {
		res, err := sm.client.SendPrompt(sm.SessionID, req)
		sm.workerDoneChan <- workerResult{Result: res, Error: err}
	}()
}
